src/credentials/providers/ec2CredentialsProvider.ts

```diff
@@ -1,27 +1,46 @@
 /*!
- * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
  * SPDX-License-Identifier: Apache-2.0
  */
 
-import * as AWS from 'aws-sdk'
-import { CredentialType } from '../../shared/telemetry/telemetry.gen'
-import { getStringHash } from '../../shared/utilities/textUtilities'
-import { CredentialsProvider, CredentialsProviderType ,CredentialsId } from './credentials'
-
+import { Credentials, EC2MetadataCredentials, MetadataService } from "aws-sdk"
+import { getLogger } from '../../shared/logger'
+import { CredentialType } from "../../shared/telemetry/telemetry.gen"
+import { CredentialsId, CredentialsProviderType } from './credentials'
+import { EnvironmentCredentialsProvider } from "./environmentCredentialsProvider"
 
 /**
  * Credentials received from EC2 metadata service.
  *
  * @see CredentialsProviderType
  */
-export class Ec2CredentialsProvider implements CredentialsProvider {
-    public constructor(private token: string) {
+export class Ec2CredentialsProvider implements EnvironmentCredentialsProvider {
+    private static readonly METADATA_SERVICE_TIMEOUT: number = 1000
+
+    private credentials: EC2MetadataCredentials | undefined
+    private region: string | undefined
+
+    public constructor(
+        private metadata: MetadataService = Ec2CredentialsProvider.makeMetadataService()
+    ) {}
+
+    public isAvailable(): Promise<boolean> {
+        return new Promise((resolve) => {
+            this.metadata.request('/latest/dynamic/instance-identity/document', (err, response) => {
+                if (response) {
+                    const document = JSON.parse(response)
+                    this.region = document['region']
+                    getLogger().verbose(`resolved instance region ${this.region} from EC2 Metadata`)
+                }
+                resolve(err ? false : true)
+            })
+        })
     }
 
     public getCredentialsId(): CredentialsId {
         return {
             credentialSource: this.getProviderType(),
-            credentialTypeId: 'default',
+            credentialTypeId: 'instance',
```

Comment:
Is it possible for the metadata service to provide multiple sets of credentials? That is, will `credentialTypeId` ever be different? The same questions apply to the other providers introduced by this PR.